### ISSUE TITLE
[@dopt/javascript] - how to initialize without React?  "Create is not a function"

### ISSUE THREAD

**Issue by jeffreybiles** [dopt/odopt#829]
I have the following code. I had to guess at the import statement, since I couldn't find one not with React in the documentation.

```
import Dopt from '@dopt/javascript';

//...

new Dopt({
  blocksAPIKey: blocksApiKey,
  userId: userApiKey,
  flowVersions: {
    'jeffrey-biles-example-flow': "uncommitted",
  }
});
```

However, when I run it, I get the error "create is not a function"

<img width="1023" alt="Screenshot 2023-08-17 at 1 24 48 PM" src="https://github.com/dopt/odopt/assets/839123/a68a696e-667a-4877-810f-46035cfccd86">

I'm using `@dopt/javascript` 3.3.0.

**Comment by jeffreybiles** [comment]
The non-simplified code is the following.  We're using EmberJS.

```js
import ENV from 'tnt-ui/config/environment';

const blocksApiKey = ENV.APP.DOPT_BLOCKS_KEY;
const userApiKey = ENV.APP.DOPT_USER_KEY;

import { tracked } from '@glimmer/tracking';
import Service from '@ember/service';
import Dopt from '@dopt/javascript';

export default class DoptService extends Service {
  @tracked dopt;

  constructor() {
    super(...arguments);

    this.dopt = new Dopt({
      blocksAPIKey: blocksApiKey,
      userId: userApiKey,
      flowVersions: {
        'jeffrey-biles-example-flow': "uncommitted",
      }
    });
  }

  get blocks() {
    return this.dopt.blocks;
  }
}
```

**Comment by kseth** [comment]
Hey @jeffreybiles, a few pointers:

```
// this has to be imported like this -- we don't default export the provider
import { Dopt } from '@dopt/javascript';
```

```
export default class DoptService extends Service {
  @tracked dopt;

  constructor() {
    super(...arguments);

    this.dopt = new Dopt({
      blocksAPIKey: blocksApiKey,

      // this should be a userId for some user which you've identified with Dopt
      // see: https://docs.dopt.com/setup/users/
      // we also have clients for doing this: https://docs.dopt.com/api/users/

      userId: "USER_ID" ,

      flowVersions: {
        'jeffrey-biles-example-flow': "uncommitted",
      }
    });
  }

  get blocks() {
    // this is a function not a getter, see: https://docs.dopt.com/sdks/javascript/classes/Dopt/#blocks
    return this.dopt.blocks();
  }
}
```

We test the JS SDK internally end-to-end in CI and are working on getting an example added to our website as we speak, sorry for the confusion!

The typedocs are generally helpful as a starting place: https://docs.dopt.com/sdks/javascript/

Let me know if I can help with anything else

**Comment by jeffreybiles** [comment]
Thanks for the quick response.

I've updated the code as suggested, and using the documentation, but the error message is the exact same as before.

Here's the new code:

```js
import ENV from 'tnt-ui/config/environment';

const blocksApiKey = ENV.APP.DOPT_BLOCKS_KEY;
const userApiKey = ENV.APP.DOPT_USER_KEY;

import { tracked } from '@glimmer/tracking';
import Service from '@ember/service';
import { Dopt } from '@dopt/javascript';
import { DoptApiClient } from '@dopt/users-javascript-browser-client';
import { task } from 'ember-concurrency';

export default class DoptService extends Service {
  @tracked dopt;
  @tracked client;

  constructor() {
    super(...arguments);

    this.load.perform();
  }

  get blocks() {
    return this.dopt?.blocks();
  }

  @task
  *load() {
    this.client = new DoptApiClient({
      apiKey: userApiKey,
    });

    const user = yield this.client.users.identifyUser({identifier: 'testing', properties: {}});

    this.dopt = new Dopt({
      blocksAPIKey: blocksApiKey,
      userId: user.identifier,
      flowVersions: {
        'jeffrey-biles-example-flow': 'uncommitted',
      },
    });
  }
}
```

**Comment by kseth** [comment]
thanks for the fast reply!

one small change (unrelated to the breakage):
```js
    this.dopt = new Dopt({
      // should be apiKey
      apiKey: blocksApiKey,
      userId: user.identifier,
      flowVersions: {
        'jeffrey-biles-example-flow': 'uncommitted',
      },
    });
```

we're separately investigating whether one of our dependencies (zustand) has an issue with esm transpilation: https://github.com/pmndrs/zustand/issues/559 -- that's probably why we don't notice this issue internally

can you share your ember / bundler configuration?

**Comment by kseth** [comment]
one hypothesis we have: in our cjs.js file, we require `zustand/vanilla` and `zustand/middleware` which are exported by `zustand` in it's package.json: https://github.com/pmndrs/zustand/blob/main/package.json#L32

it's feasible that your bundler doesn't capture these exports -- for example, webpack only solved this in v5 (https://github.com/webpack/webpack/issues/9509)

**Comment by jeffreybiles** [comment]
We've specified `"webpack": "^5.80.0"` in our `package.json`.  Our lockfile does include webpack 4 as well, but I think it's webpack 5 that gets used.

Ember handles most of the bundling for us.  Here's the relevant parts of the `ember-cli-build.js` file:

```js
autoImport: {
  webpack: {
    module: {
      rules: [fileLoaderConfig],
    },
    plugins: [
      new MomentLocalesPlugin({
        // 'en' is built into moment and cannot be removed. This strips the others.
        localesToKeep: [],
      }),
    ],
  },
},
```

**Comment by jeffreybiles** [comment]
I also tried adding `"zustand": "^4.4.1"` to the `package.json`, but it didn't seem to change anything

**Comment by jeffreybiles** [comment]
I noticed this node_modules path for the vanilla Javascript package includes the a reference to React 16.4.0:

```
./node_modules/.pnpm/@dopt+javascript@3.3.0_react@16.14.0/node_modules/@dopt/javascript/dist/index.cjs.js
```

Is that intentional?

**Comment by joemckenney** [comment]
Hey @jeffreybiles, thanks for the info about your configuration. I'm going to try to set up a reproduction of this issue locally now.  I will update here as we learn more.

**Comment by kseth** [comment]
@jeffreybiles -- one last suggestion as we're getting a repro setup, could you try adding the following to your webpack config?

```
  //...
  resolve: {
    conditionNames: ['default'],
  }
```

According to webpack, this is how you get webpack to understand the `exports` from a package.json: https://webpack.js.org/configuration/resolve/#resolveconditionnames

**Comment by jeffreybiles** [comment]
Thanks for reproducing!

Adding the above code gives me this error

<img width="593" alt="Screenshot 2023-08-17 at 7 07 12 PM" src="https://github.com/dopt/odopt/assets/839123/f5c29a50-df0b-4ddc-98af-7a64f526a823">

Unfortunately, Ember sometimes makes its own rules in an attempt to make things easier, so things don't port over 1:1

I probably won't have time to work on this much in the next couple days due to shifting business priorities, but will be sure to check in every once in a while to try any suggestions

**Comment by joemckenney** [comment]
This [repository](https://github.com/dopt/ember-js-with-dopt-javascript) reproduces the issue and [this commit](https://github.com/dopt/ember-js-with-dopt-javascript/commit/fc9a01578f99d86b4efeda7d81a7f8c1f9897b34) fixes it.   Let me know if that works for your configuration.

**Comment by kseth** [comment]
We've also fixed the `zustand` CJS bundling issue in the most recent version of @dopt/javascript (3.4.0) so you shouldn't need to workaround @joemckenney describes above.

I updated the repository to this [commit](https://github.com/dopt/ember-js-with-dopt-javascript/commit/60e0fa076d43ebe2584dcc2ac21df408d21afa65) and verified it works as expected.

Going to close this issue, and thank you for reproting @jeffreybiles!
